bindings/py/cpp_src/bindings/algorithms/py_Connections.cpp

```diff
@@ -0,0 +1,156 @@
+/* ---------------------------------------------------------------------
+ * Numenta Platform for Intelligent Computing (NuPIC)
+ * Copyright (C) 2019, David McDougall
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses.
+ * --------------------------------------------------------------------- */
+
+/** @file
+ * PyBind11 bindings for Connections class
+ */
+
+#include <bindings/suppress_register.hpp>  //include before pybind11.h
+#include <pybind11/pybind11.h>
+#include <pybind11/iostream.h>
+#include <pybind11/numpy.h>
+#include <pybind11/stl.h>
+
+#include <string>
+#include <sstream>
+
+#include <nupic/algorithms/Connections.hpp>
+
+namespace py = pybind11;
+using namespace nupic;
+using namespace nupic::algorithms::connections;
+using nupic::sdr::SDR;
+
+namespace nupic_ext
+{
+  void init_Connections(py::module& m)
+  {
+    py::class_<Connections> py_Connections(m, "Connections",
+R"(Compatibility Warning: This classes API is unstable and may change without warning.)");
```

Comment:
:+1: thanks

Reply:
I didn't like adding this note, because I fear it will scare off users.

However you're probably correct that this API is unstable.  I already thought of several things which I'd like to change about this API:
- I'd like to wrap the `SegmentData` & `SynapseData` classes, and for those wrappers to contain pointers back to the `Connections` object.  This way python users can treat segments and synapses as objects instead of the current: as integer handles into a connections object.
    * Example current API: `connections.updateSynapsePermanence(synIdx, newPerm)`
    * Example desired API: `synapse.updatePermanence(newPerm)` which inside of the pybinding code will call the current API.
- I'd like to provide Numpy array interfaces for methods which are typically called every compute cycle, with the goal of python calling them once per compute cycle.  Currently these need to be called once per segment, they should be called once with an array of segments.
   * adaptSegment
   * raisePermanencesToThreshold
   * bumpSegment

Reply:
And for clarity: these are ideas for the future, not this current PR.  If/when this PR merges I will move these ideas into the issue #243.

Reply:
> 'd like to provide Numpy array interfaces for methods which are typically called every compute cycle, with the goal of python calling them once per compute cycle. Currently these need to be called once per segment, they should be called once with an array of segments.  

I would like the change even in c++, for the low-level methods to take (sparse) vector, instead of a single integer (index). This would move the loops to a higher code. //for later